### ISSUE TITLE
[Not a bug] PSA: if you are seeing locks in your DB...

### ISSUE THREAD

**Issue by maxkwallace** [flask-admin/flask-admin#2008]
Not an issue or a bug report, just want this to be searchable in case anyone else does the same (silly) thing in the future.

- - -

We have a Flask app (1.1.2) running on Python 3.6 and we're using flask-admin 1.5.6. We're using SQLAlchemy, and Alembic for DB migrations. When we ran our equivalent of `alembic upgrade` (using [Flask-Migrate](https://flask-migrate.readthedocs.io/en/latest/)) and our web server was already running, the migration would sometimes hang indefinitely just before finishing. We could only "un-hang" it by killing the web server or restarting the database.

One could replicate this locking in development, for example, with a simple migration to add the column `bar` to the table `foo`. It looks like this:

```
(venv) ~/directory$ python3 [flask-migrate wrapper script] db upgrade
11:27:04,864 [INFO] sqlalchemy - select version()
11:27:04,864 [INFO] sqlalchemy - {}
11:27:04,865 [INFO] sqlalchemy - select current_schema()
11:27:04,865 [INFO] sqlalchemy - {}
11:27:04,866 [INFO] sqlalchemy - SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
11:27:04,866 [INFO] sqlalchemy - {}
11:27:04,867 [INFO] sqlalchemy - SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
11:27:04,867 [INFO] sqlalchemy - {}
11:27:04,868 [INFO] sqlalchemy - show standard_conforming_strings
11:27:04,868 [INFO] sqlalchemy - {}
11:27:04,869 [INFO] alembic.runtime.migration - Context impl PostgresqlImpl.
11:27:04,869 [INFO] alembic.runtime.migration - Will assume transactional DDL.
11:27:04,870 [INFO] sqlalchemy - BEGIN (implicit)
11:27:04,870 [INFO] sqlalchemy - select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=%(name)s
11:27:04,870 [INFO] sqlalchemy - {'name': 'alembic_version'}
11:27:04,871 [INFO] sqlalchemy - SELECT alembic_version.version_num 
FROM alembic_version
11:27:04,871 [INFO] sqlalchemy - {}
11:27:04,882 [INFO] alembic.runtime.migration - Running upgrade dc1bb532ad90 -> 9bfab3de489d
11:27:04,883 [INFO] sqlalchemy - ALTER TABLE foo ADD COLUMN bar VARCHAR
11:27:04,883 [INFO] sqlalchemy - {}
[hangs forever]
```

Following the example queries on https://wiki.postgresql.org/wiki/Lock_Monitoring , when the migration locks, one sees the following in the database:

```
blocked_statement | current_statement_in_blocking_process
ALTER TABLE foo ADD COLUMN bar VARCHAR | SELECT foo.external_id AS foo_external_id, foo.created_date AS foo_created_date ...
ALTER TABLE foo ADD COLUMN bar VARCHAR | SELECT foo.external_id AS foo_external_id, foo.created_date AS company_created_date ...
```

(I'm not sure why there are two identical rows.) 

The `SELECT foo.external_id ...` statements on the right match up with the logs from the server:

```
(venv) /directory$ python3 app.py
11:27:02,903 [INFO] sqlalchemy - select version()
11:27:02,903 [INFO] sqlalchemy - {}
11:27:02,904 [INFO] sqlalchemy - select current_schema()
11:27:02,904 [INFO] sqlalchemy - {}
11:27:02,905 [INFO] sqlalchemy - SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
11:27:02,905 [INFO] sqlalchemy - {}
11:27:02,906 [INFO] sqlalchemy - SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
11:27:02,906 [INFO] sqlalchemy - {}
11:27:02,906 [INFO] sqlalchemy - show standard_conforming_strings
11:27:02,906 [INFO] sqlalchemy - {}
11:27:02,907 [INFO] sqlalchemy - BEGIN (implicit)
11:27:02,908 [INFO] sqlalchemy - SELECT foo.external_id AS foo_external_id, foo.created_date AS foo_created_date, foo.foo_id AS foo_foo_id, foo.name AS foo_name, 
FROM foo
11:27:02,909 [INFO] sqlalchemy - {}
...

 * Serving Flask app "project" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
11:27:02,959 [INFO] werkzeug on http://0.0.0.0:5000/ (Press CTRL+C to quit)
```

And that `SELECT foo.external_id ...` query appears to be coming from flask-admin. If I disable flask-admin completely in the server process, I cannot replicate the locking behavior.

- - - 

### Solution:

We have the following custom filter defined:

```
class RelationshipFilter(BaseSQLAFilter):
    def __init__(self, cls, target_cls, session, relationship_method_name=None):
        name = target_cls.__name__

        # cls is the class with the relationship, target_cls is the class the
        # relationship references.  It's important to note that if `cls` is
        # not the same as the view class, Flask will attempt to figure out
        # how to do a join.
        # self.name is the label displayed on the dropdown component / filters
        self.cls = cls
        self.target_cls = target_cls
        self.name = name_override or name
        self.method_name = (relationship_method_name or
            inflection.underscore(name)) + '_id'
        self.session = session
        self.column = getattr(cls, self.method_name)

        # Don't know what this is used for. It may not be needed.
        self.key_name = None

        # If self.data_type is None and the Filter has an options property,
        # flask-admin/flask_admin/static/admin/js/filters.js will assume it
        # should render a select dropdown of options.
        self.data_type = None

    # Flask-Admin expects a list of tuples as the return value of this method.
    # The first value in the tuple is what gets passed to 'apply' and the second
    # is the string that appears in the options dropdown.
    @property
    def options(self):
        models = [
            (m.id, repr(m)) for m in self.session.query(self.target_cls).all()]
        return sorted(models, key=lambda m: m[1])

    # Not sure if this is needed.
    def get_options(self, view):
        return self.options

    def apply(self, query, value):
        return query.filter_by(**{ self.method_name: value })

    def operation(self):
        return f'{self.name} is'
```

This filter is extremely useful for filtering views based on generic SQLAlchemy relationships. What I didn't realize is the `options` property is called when the view(s) are first initialized (not in the context of a request), and for some reason I don't understand, the (implicit) transaction SQLAlchemy creates around the query `self.session.query(self.target_cls).all()` in `options` will stay open and lock up the table.

The code in `options` should be changed to:

```
@property
def options(self):
    self.session.rollback()
    models = [
        (m.id, repr(m)) for m in self.session.query(self.target_cls).all()]
    self.session.rollback()
    return sorted(models, key=lambda m: m[1])
```

tl;dr `Filter.options` is called on view initialization, not on a per request basis, so if you do any queries inside a custom filter, make sure you roll back the transaction when you're done!

**Comment by maxkwallace** [comment]
Fix is in the OP
